Feat: Main 페이지 완성 및 Detail 페이지 UI layout 완성

JISU-Y opened the pull request:
Main 페이지
- [x] ToolTip 추가 및 적용
- [x] Grid 적용

Detail 페이지
- [x] Image, Review, Info section 나눔, data fetching 완료
- [x] Button hover 효과 같은 요소들 컴포넌트화
- [x] Image 슬라이드 완료(Initial Slide 설정 추가)
- [x] 오른쪽 제품 Info 부분 완료
- [x] 아래쪽 Review 부분 완료

JISU-Y committed:
Refactor: Main 페이지 작업
Grid 적용 및 svg icon 삽입

JISU-Y committed:
Feat: ToolTip Template 추가

JISU-Y committed:
Feat: Main 페이지 완성
ToolTip 추가 및 Grid 적용

JISU-Y committed:
Feat: Detail 페이지 UI layout 완성
Image, Review, Info section 나눔, data fetching 완료

JISU-Y committed:
Detail 페이지 작업
Button hover효과 같은 요소들 컴포넌트화
Image 슬라이드, 오른쪽 제품 Info, 아래쪽 Review 부분 완료

addiescode-sj commented on src/components/BgAnimatedButton/BgAnimatedButton.tsx:
```diff
@@ -0,0 +1,85 @@
+  buttonLabel,
+  useToggle,
+  shouldShow,
+  setShouldShow,
```
setState 액션을 자식컴포넌트로 넘기는 건 대표적인 안티패턴중에 하나입니다~! 이유는 여러가지가 있을 수 있겠지만 제일 큰 문제는 자식컴포넌트에서 setState액션이 실행될때와 부모컴포넌트에서 setShouldShow액션이 실행될때 충돌이 일어날수있기때문인데요.  그래서 setState를 넘기지 마시고, 버튼을 클릭함으로써 shouldShow의 값이 토글되게끔 만들어야하는거라면 이런 방식을 대신 생각해볼수 있을것같아요. 

1. setShouldShow prop 제거, 부모컴포넌트에서 setShouldShow 액션이 포함된 핸들러 함수 작성
2. 해당 핸들러 함수의 인자(parameter)에 토글되는 boolean값 전달
3. 핸들러 함수를 자식 컴포넌트에 props로 넘기고, 자식 컴포넌트 자체에서 shouldshow state를 따로 관리하며 boolean 체크
4. 자식 컴포넌트 Buttoin onClick 함수에서 shouldshow로 관리하는 state값을 인자로 넘김. 

이렇게 공유하는 값을 인자로 가지는 핸들러 함수를 하나 만들고 부모,자식의 고유의 state를 따로 관리해 분리하면 부모-자식간에 안전하게 데이터 공유를 할 수 있는 구조가 됩니다.

시도해보세요~! :)

JISU-Y replied:
궁금한 것이 있습니다!

자식 컴포넌트 자체에서 shouldshow state(boolean)을 체크할 때 button onClick으로 하면 부모에게 전달받은 handler 함수도 그 button의 onClick에 넣어주어야 하는데 어떻게 해야하나요..?

onClick에는 자식 컴포넌트에서 관리하는 setState만 사용하고, 그 state가 변화할 때 useEffect로 부모 핸들러에 state값을 전달해주어야 하는건가요?

      const [toggle, setToggle] = useState(false);
  
      const handleToggle = () => setToggle(prev => !prev);
    
      useEffect(() => {
        if (!toggleShow) return;
        toggleShow(toggle);
      }, [toggleShow, toggle]);
    
      return (
        <Button textColor={textColor} bgColor={bgColor} onClick={handleToggle}>

아니면 따로 자식 컴포넌트에 버튼 Wrapper div를 만들어서 onClick 함수를 분리해야하는지 잘 모르겠습니다!

      const [toggle, setToggle] = useState(false);
  
      return (
        <div onClick={() => setToggle(prev => !prev)}>
          <Button
            textColor={textColor}
            bgColor={bgColor}
            onClick={() => toggleShow!!(!toggle)}
          >

addiescode-sj replied:
아하 네 부모-자식간에 함수로 데이터를 공유하는 패턴에 대해 이번주 페어프로그래밍 시간에 알려드리면 좋을것같네요.
자세하게는 페어프로그래밍으로 리팩토링하면서 진행해드릴게요 :) 

우선 제가 말씀드린 구조는 아래와 같습니다.

부모 컴포넌트와 자식컴포넌트가 핸들러 함수를 공유하는 형태이고, 
해당 핸들러 함수에는 부모컴포넌트에서 state를 관리하고있는 값을 넘겨줍니다.

또, 자식컴포넌트에서 토글을 state값으로 따로 관리해야하는데 이때 state의 초기값으로 부모컴포넌트에서 관리하고있는 토글값이 필요하다면 해당 값을 부모컴포넌트에서 prop으로 넘겨줍니다. 

이런식으로 자식컴포넌트에서 state 초기값을 부모에게서 넘겨받은 state로 넣어주고,
onClick 핸들러함수를 만들어서 자식 컴포넌트에서 관리하고있는 state값을 setState해주고 해당 값을
부모 컴포넌트에서 넘겨받은 핸들러로 넘겨주면 값이 공유되겠죠? ㅎㅎ

```tsx
 // 부모 컴포넌트

  const [toggle, setToggle] = useState(false);

 const handleToggle = (toggleValue: boolean) => {
   setToggle(toggleValue)
 }

```

```tsx

 // 자식 컴포넌트
 
  const [toggle, setToggle] = useState(부모컴포넌트에서 넘겨받은 props- state 초기값);

 const handleOnClick = () => {

   setToggle(!toggle);

    // 부모에게 넘겨받은 함수
    handleToggle(toggle);

}

  <Button
        textColor={textColor}
        bgColor={bgColor}
        onClick={handleToggle}
   >

```

addiescode-sj commented on src/components/BgAnimatedButton/BgAnimatedButton.tsx:
```diff
@@ -0,0 +1,85 @@
+  bgColor?: string;
+}>`
+  ${({ textColor, bgColor }) => css`
+    font-size: 13px;
```
styles 선언하는 순서에 대해서도 조금 코멘트 드릴게요.

```css
 /* Positioning */
  position: relative;

  /* Display & Box Model */
  display: flex;
  align-items: center;
  overflow: hidden;
  box-sizing: border-box;
  padding: 12px 15px;
  border: none;

  /* Color */
  background: none;
  color: ${textColor ?? COLORS.mainFont};
  
  /* Text */
  font-size: 13px;
  font-weight: 500;

  /* Other */
  cursor: pointer;

  &::after {
      content: '';
      position: absolute;
      width: 100%;
      height: 100%;
      top: 0;
      left: 0;
      border-radius: 30px;
      transform: scale(0.8);
      background: ${bgColor ? bgColor : COLORS.hoverGray};
      opacity: 0;
      transition: ${TRANSITION.quick};
    }
    
   &:hover::after {
      transform: scale(1);
      opacity: 1;
    }

```

이런식으로 css order를 특성별로 묶어서 작성해주시면 훨씬 리뷰어가 읽기 수월해질거예요.
[리소스](https://css-tricks.com/poll-results-how-do-you-order-your-css-properties/) 읽어보시는 것을 추천드립니다!

JISU-Y replied:
어렴풋이 본 기억은 있는데.. 제대로 몰라서 적용 못하고 구현만 하고 있었네요ㅠㅠ

모두 수정해두겠습니다! 감사합니다

addiescode-sj commented on src/components/TabMenu/TabMenu.tsx:
```diff
@@ -1,26 +1,38 @@
+function TabMenu({
+  list,
+  currentTab,
+  setCurrentTab,
```
currentTab은 굳이 부모 컴포넌트에서 관리될 필요가 없는 state같아요. 자식컴포넌트에서 관리하게끔 바꿔주세요~!

JISU-Y replied:
Detail 페이지에서 ReviewSection 컴포넌트에서 사용하는 TabMenu는 보여줄 컨텐츠가 따로 없어서 currentTab을 부모에서 따로 사용하지 않아도 될 것 같습니다!

그런데 Main 페이지에서 UniqueList에서는 tab Menu 클릭될 때마다 currentTab으로 컨텐츠들을 filtering하고 있어서 부모에서도 currentTab이 지금 어떤 값인지 알고 있어야 할 것 같습니다..!

위에서 리뷰 주신 것처럼 setState로 넘기지 않고 핸들러 함수 만들어서 Props로 넘기도록 수정해도 괜찮을까요??

addiescode-sj replied:
넵~! 그럼 그렇게 수정하시면 좋을 것 같네요 :)

addiescode-sj commented on src/pages/Detail/sections/ReviewSection/ReviewSection.tsx:
```diff
@@ -0,0 +1,35 @@
+import * as S from './ReviewSection.style';
+
+function ReviewSection({ reviewCount }: { reviewCount: number }) {
+  const [currentTab, setCurrentTab] = useState(0);
```
currentTab이 부모 컴포넌트에서 관리될 필요가 없는 state값인것같은데 맞나요~?
이런 경우 굳이 위에서 리뷰드린 내용처럼 함수를 통해 데이터를 전달하지않고 자식 컴포넌트에서만 state관리를 해주시면 좋을것같습니다.

addiescode-sj commented on src/pages/Detail/sections/SideInformation/SideInformation.style.ts:
```diff
@@ -0,0 +1,165 @@
+  ${({ isOverLimit }) =>
+    isOverLimit &&
+    css`
+      height: auto;
```
height: auto를 빼면 동작이 달라지나요? 
auto는 기본적으로 들어가있는 속성이라 빼주셔도 됩니다 :)

addiescode-sj reviewed (changes requested):
수고하셨습니다! 👍 

이번주는 확실히 제가 리뷰드린 내용을 잘 숙지하시고 활용하고계셔서 리뷰 양이 확 줄었네요.

코드상으로 중요하게 고쳐야하는 부분들 (state관리, 부모-자식간 데이터공유에 대한 리뷰) + 스타일속성 순서 정리하는 팁 위주로 보시고 반영하시면 좋을 것 같습니다. 

[노션에서 질문주신 내용](https://addiescodemade.notion.site/0bbe42bf595d41f09835788b77271e23)에 답변드리자면
사실 grid-template은 IE를 아직 많이 쓰는 우리나라에서는 사용을 지양하는걸로 많이 알고있는데,
지금은 IE11까지도 잘 지원하고있는, 즉 production에서 충분히 활용할 수 있는 속성이라서 쓰셔도됩니다. ([참고](https://caniuse.com/css-grid))

지수님이 grid-template-areas로 스타일링하는게 직관적이라고 생각하신이유에 대해 조금 궁금합니다! 
저는 컬럼/열마다 조금 더 복잡한 스타일링을 써내려가기엔 grid를 활용하는게 좋다고 생각하는 편입니다 ㅎㅎ

하지만 지금 보면, 결과물 자체는 같은 비율의 row가 두개 생기고, 같은 비율의 column이 6개가 생기는거니까 그리드를 활용하는것보다는 flexbox를 div 두개에 나눠서 두번 활용하고 각각의 flex-item이 동일한 비율로 차지할수있게끔 스타일링하는게 더 간단해보이지않을까요? 이렇게 해주면 인덱스값을 신경쓰지않아도되고, 굳이 클래스네임을 붙여 관리할 필요도 없어지니까요 :)

JISU-Y commented:
처음에 flex를 사용해서 하려다가 grid로 한 이유는

flex로 하면 div를 두 개 생성해야 하는데, 똑같은 div를 두번 써야하지 않을까 생각되었고, div를 나누게 되면 tabProductList를 0-2번째 3-5번째로 나누어서 각각 div에 담아야 하는 것인지 구현 방법이 잘 생각나지 않았습니다ㅠㅠ!

그래서 jsx 요소는 최대한 그대로 두고 index를 이용해서 grid로 구현하는 것이 (grid-template-areas을 이용하면 일단 각각의 크기 설정은 쉬우니) 쉽지 않을까 생각해서 구현하게 되었습니다

추가로 궁금한 것이, styled component 사용할 때 className을 사용하는 것은 지양해야 하는 패턴인가요??


지금 flex로 구현해보려고 하는데, 정확히 어떻게 해야하는 것인지 감이 잘 안잡힙니다ㅠㅠ

map으로 list div가 나란히 6개가 있는데, 이것을 div로 남아서 담으면 slice를 하는 방법 밖에 생각이 나질 않습니다..

```
      <S.TabContents>
        <S.ImageCardWrapper>
          {tabProductList?.slice(0, 3).map(({ imageUrl, price }) => (
            <S.ImageCardBox key={imageUrl}>
              <ImageCard
                width="100%"
                height="100%"
                price={price}
                image={imageUrl}
              />
            </S.ImageCardBox>
          ))}
        </S.ImageCardWrapper>
        <S.ImageCardWrapper>
          {tabProductList?.slice(3, 6).map(({ imageUrl, price }) => (
            <S.ImageCardBox key={imageUrl}>
              <ImageCard
                key={imageUrl}
                width="100%"
                height="100%"
                price={price}
                image={imageUrl}
              />
            </S.ImageCardBox>
          ))}
        </S.ImageCardWrapper>
      </S.TabContents>
```

기본적으로 map을 돌려서 나열하면 이렇게 되는 것인데, 
```
<S.TabContents>
    <S.ImageCardWrapper>
        <ImageCard />
        <ImageCard />
        <ImageCard />
        <ImageCard />
        <ImageCard />
        <ImageCard />
    </S.ImageCardWrapper>
</S.TabContents>
```

어떻게 이 위의 imageCard들을 아래와 같이 Div에 담아서 flex로 처리할 수 있을까요..?

```
<S.TabContents>
    <S.ImageCardWrapper>
        <Div>
            <ImageCard />
            <Div>
                <ImageCard />
                <ImageCard />
            </Div>
        </Div>
        <Div>
            <ImageCard />
            <Div>
                <ImageCard />
                <ImageCard />
            </Div>
        </Div>
    </S.ImageCardWrapper>
</S.TabContents>
```

JISU-Y committed:
refactor: tabMenu 및 BgAnimatedButton 수정
setState Props에서 제거

JISU-Y committed:
refactor: css attribute 순서 리팩토링

JISU-Y committed:
feat: Cart 페이지 작업
Cart 페이지 생성
Cart 페이지 Product Info 왼쪽 섹션 마크업
Router 추가

addiescode-sj commented:
앗, Flex item에서 또 display: flex로 flex를 중첩해서 선언할수있어요.
className을 붙인다고 성능이 떨어지거나하는건 전혀 아니지만, 각각의 index에 맞게 네이밍을 해주고 해당 네이밍에 맞는 스타일을 따로 선언해주는 형태가 간단하게 flex 중첩해서 해결될 스타일링이라면 후자의 경우가 더 낫다는 의미에서 의견드려봤습니다 ㅎㅎ

말씀주신 부분은 제가 Pull 받아서 스타일링해보고, 다시 말씀드릴게요~! 
우선 위에서 말씀드린 리뷰들 적용하는 시간 갖고 계시면 좋을것같습니다 ㅎㅎ


**(+ 추가 내용)**

- undefined인 상태였을때 체크해서 함수를 어떻게 실행하는지에 대해서도 궁금하셨던거군요! 

간단합니다. 마찬가지로 [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)을 이용해주시면 됩니다. 

```tsx
  const handleToggle = () => {
    setToggle(!toggle);
    toggleShow?.(toggle);
  };
```

- 저는 해당 ui를 보니까 왼쪽엔 큰 카드형식 하나, 오른쪽에는 작은 카드형식 두개 이렇게 flexbox를 구성하고 해당 컴포넌트를 인덱스에 맞춰 두번 반복하는 형태가 떠올랐어요. Etsy hompage를 inspect해보니, Flex-container안에서 Flex-box두개를 구성하고 해당 flex-box에서 또 flex를 해주는 방식으로 구현된것같네요. flex-basis라는 속성을 어떻게쓰는건지, 지수님 방법도 좋지만 이런건 어떻게 컴포넌트화시키는게 좋은지 페어프로그래밍때 또 한번 알려드려볼게요! :) 

<img width="702" alt="image" src="https://user-images.githubusercontent.com/69444278/154553121-20f9b6a6-a263-4fec-bdc7-bdefd50e2143.png">

JISU-Y committed:
feat: Cart 페이지 작업
Cart 페이지 오른쪽 Payment 섹션 마크업

JISU-Y committed:
refactor: tabMenu 및 BgAnimatedButton 수정
state 자식에서 따로 관리하여 부모 컴포넌트와 state 공유하도록 변경

JISU-Y committed:
help: UniqueList 컴포넌트 flex

JISU-Y committed:
feat: Cart 페이지 작업
Cart 페이지 마크업 및 스타일 적용 완료

JISU-Y committed:
feat: Cart 페이지 작업 완료
장바구니 아이템 추가, 수량변경, 삭제 기능 추가
타이틀 클릭 시 디테일 페이지 이동, keep shopping 클릭 시 메인 페이지 이동 구현

JISU-Y committed:
feat: Header 장바구니 버튼 작업
장바구니 버튼 오른쪽 상단 개수 표시 추가

JISU-Y committed:
refactor: minor change
Header Logo 클릭 시 main으로 이동
Cart 페이지 명세서 계산 오류 개선

JISU-Y committed:
Merge branch 'feat/cartPage' into feat/detailPage

JISU-Y committed:
refactor: Cart 페이지 작업 merge

addiescode-sj committed:
refactor: appcontext, form control, naming, type

addiescode-sj committed:
refacotr: errors -> dirtyFields

JISU-Y committed:
refactor: 주석 정리, minor change
